src/backend/vulkan/src/device.rs

```diff
@@ -356,6 +800,85 @@ impl d::Device<B> for Device {
         }
     }
 
+    unsafe fn create_graphics_pipeline<'a>(
+        &self,
+        desc: &pso::GraphicsPipelineDesc<'a, B>,
+        cache: Option<&n::PipelineCache>,
+    ) -> Result<n::GraphicsPipeline, pso::CreationError> {
+        debug!("create_graphics_pipeline {:?}", desc);
+
+        let buf = GraphicsPipelineInfoBuf::new(self, desc);
```

Comment:
I'll try to get this out of the Box so that everything (beside the Vecs) is allocated in the stack on that particular function.
As for the Vecs allocations, they depend on a function parameter, so the size may change between function calls on the same device (and, if this function is being called more than once, they probably will, which makes a buffer need to be resized and waste memory when not dealing with the bigger sizes).
Also, having a buffer for that breaks the "you pay for what you ask only" thing. Not all programs may want to create graphical pipelines.

Reply:
So after moving this out of the box there is going to be a few `Vec` things left. I don't think there is any concern about memory waste here (since we don't expect those vectors to be taking any significant amount of memory), and I don't think it breaks the "you pay for what you ask only" thing. It's simply better to not heap-allocate (and re-use) instead of allocating.

The problem is in the details - implementation ones. Device is `&self` here, so if we want to re-use any mutable structures, we'll need to do something smart: have an internally mutable collection of those buffers, where we request one that isn't locked (or create a new one), use it, and then unlock it. Given the complexity, we don't have to bother with it right away - your PR is already a nice improvement.